api/src/main/java/org/apache/iceberg/actions/RemoveOrphanFiles.java

```diff
@@ -0,0 +1,79 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.function.Consumer;
+
+/**
+ * An action that removes orphan files in a table.
+ * <p>
+ * A metadata or data file is considered orphan if it is not reachable by any valid snapshot.
+ * The set of actual files is built by listing the underlying storage which makes this operation
+ * expensive.
+ */
+public interface RemoveOrphanFiles extends Action<RemoveOrphanFiles.Result> {
+  /**
+   * Passes a location which should be scanned for orphan files.
+   * <p>
+   * If not set, the root table location will be scanned potentially removing both orphan data and
+   * metadata files.
+   *
+   * @param location the location where to look for orphan files
+   * @return this for method chaining
+   */
+  RemoveOrphanFiles location(String location);
+
+  /**
+   * Removes orphan files only if they are older than the given timestamp.
+   * <p>
+   * This is a safety measure to avoid removing files that are being added to the table.
+   * For example, there may be a concurrent operation adding new files while the action searches
+   * for orphan files. New files may not be referenced by the metadata yet but they are not orphan.
+   * <p>
+   * If not set, defaults to a timestamp 3 days ago.
+   *
+   * @param olderThanTimestamp a long timestamp, as returned by {@link System#currentTimeMillis()}
+   * @return this for method chaining
+   */
+  RemoveOrphanFiles olderThan(long olderThanTimestamp);
+
+  /**
+   * Passes an alternative delete implementation that will be used for orphan files.
+   * <p>
+   * This method allows to customize the delete func. For example, one may set a custom delete func
+   * and collect all orphan files into a set instead of physically removing them.
+   * <p>
+   * If not called, uses the table's {@link org.apache.iceberg.io.FileIO io} implementation.
+   *
+   * @param deleteFunc a function that will be called to delete files
+   * @return this for method chaining
+   */
+  RemoveOrphanFiles deleteWith(Consumer<String> deleteFunc);
+
+  /**
+   * The action result that contains a summary of the execution.
+   */
+  interface Result {
+    /**
+     * Returns locations of orphan files.
+     */
+    Iterable<String> orphanFileLocations();
```

Comment:
I am using `Iterable` vs some more specific implementations. Let me know what everybody thinks.

Reply:
Is it possible that the orphaned files could be generated lazily? Would a `Stream` be more appropriate?

Reply:
Lazy could make sense, [and some impls of Iterable could be lazy], but in this case, who owns responsibility releasing the underlying  resource?

Reply:
Closeable iterable then?

Reply:
I think `CloseableIterable` may make sense. We use it mostly in `TableScan`. The only downside that it will complicate the usage a bit. Some of those iterables can be large so I'd want to avoid extra copies. Let me think this through.

Reply:
Well, I think it makes sense to use `ClosableIterable` in this particular case. At the same time, other places in this PR could continue to use `Iterable`.

Let me know what you think, @rymurr @rdsr @RussellSpitzer.

Reply:
I was looking at the V1 Action impl for RemoveOrphanFiles. Seems like the orphan file list is computed upfornt. @aokolnychyi do you see the list being lazy in this particular case? My argument was mostly for lazy streams (or lazy impl of `Iterables`). If all our results are computed eagerly then returning a List is fine as that would remove the question of resource cleanup and ownership. I'm fine with `Iterables` also]

Reply:
We don't have a use case where this would be evaluated lazily yet. I can make it `ClosableIterable` if we think other implementations may benefit from this but it will definitely make using this API harder.

My motivation for changing the return type to `Iterable` is that it allows us to change the underlying representation. For example, we could migrate from `List` to `Set` in the internal implementation.

Reply:
> Well, I think it makes sense to use `ClosableIterable` in this particular case. At the same time, other places in this PR could continue to use `Iterable` as it is unlikely they will benefit from lazy materialization.

Why? Cases like `Iterable<ManifestFile> rewrittenManifests()` can also benefit from using `CloseableIterable`. I feel if we want to do it for one, we should do it for all, because we cannot just assume one result might need to close resource, the other does not. If an engine does it for one action, it is likely that it will have this behavior for all actions.

With that being said, I agree that using `CloseableIterable` would make the interface harder to use. Why not just use `Iterable`, and let `Result` be the object to implement `close()`? If one engine has a closable iterable, it can be closed there.

Reply:
> Cases like Iterable<ManifestFile> rewrittenManifests() can also benefit from using CloseableIterable. I feel if we want to do it for one, we should do it for all, because we cannot just assume one result might need to close resource, the other does not. If an engine does it for one action, it is likely that it will have this behavior for all actions.

Given that it adds complexity, I feel like this should be considered on a case by case basis. For orphan files, it may make sense as we can stream the result but I am not sure about cases like `rewrittenManifests`. Most likely, we would need to materialize the list of manifests to rewrite at some point as we need to process all entries in them.

> Why not just use Iterable, and let Result be the object to implement close()?

I think this is a better idea than using `ClosebleIterable`. In this case, we probably should make all action results closable for consistency.

What does everybody on this thread think?

Reply:
+1

Reply:
There is also an option of having engine-specific return types. However, that is going to complicate things further. It means we need engine-specific actions and to propagate `ThisT` everywhere in base actions. Plus, the return type may be substantially different from engine to engine.

Reply:
Do we need this much complication? I don't think that we would necessarily want to make this lazily evaluated because we expect the set of orphan files to be small and a concrete type is easiest for a caller to handle. I have no problem with this being a `List`, but `Iterable` is also fine with me to be able to change it to a `Set` later.

Reply:
Well, it is hard for me to predict how useful that would be for other engines. We don't really use it in Spark and maybe if we treat these results as summaries, that should be fine. In the worst case, we can cast action to an engine specific type and call a method that returns a custom object. For example, we have `expire` that returns `DataFrame` in Spark.